src/wolfio.c

```diff
@@ -2714,4 +2714,385 @@ int wolfSSL_SetIO_LwIP(WOLFSSL* ssl, void* pcb,
+            ret = isotp_receive_flow_control(ctx);
+        }
+        /* Frame delay <= 0x7f is in ms, 0xfX is X * 100 us */
+        if (ctx->frame_delay) {
+            if (ctx->frame_delay <= ISOTP_MAX_MS_FRAME_DELAY) {
+                ctx->delay_fn(ctx->frame_delay * 1000);
+            } else {
+                ctx->delay_fn((ctx->frame_delay & 0xf) * 100);
+            }
+        }
+        switch (ret) {
+            /* Clear to send */
+            case ISOTP_FLOW_CONTROL_CTS:
+                if (isotp_send_consecutive_frame(ctx) < 0) {
+                    WOLFSSL_MSG("ISO-TP error sending consecutive frame");
+                    return WOLFSSL_CBIO_ERR_GENERAL;
+                }
+                break;
+            /* Receiver says "WAIT", so we wait for another flow control
+             * packet, or abort if we have waited too long */
+            case ISOTP_FLOW_CONTROL_WAIT:
+                ctx->wait_counter += 1;
+                if (ctx->wait_counter > ISOTP_DEFAULT_WAIT_COUNT) {
+                    WOLFSSL_MSG("ISO-TP receiver told us to wait too many"
+                            " times");
+                    return WOLFSSL_CBIO_ERR_WANT_WRITE;
+                }
+                break;
+            /* Receiver is not ready to receive packet, so abort */
+            case ISOTP_FLOW_CONTROL_ABORT:
+                WOLFSSL_MSG("ISO-TP receiver aborted transmission");
+                return WOLFSSL_CBIO_ERR_WANT_WRITE;
+            default:
+                WOLFSSL_MSG("ISO-TP got unexpected flow control packet");
+                return WOLFSSL_CBIO_ERR_GENERAL;
+        }
+    }
+    return 0;
+}
+
+int ISOTP_Send(WOLFSSL* ssl, char* buf, int sz, void* ctx)
+{
+    int ret;
+    struct isotp_wolfssl_ctx *isotp_ctx;
+    (void) ssl;
+
+    if (!ctx) {
+        WOLFSSL_MSG("ISO-TP requires wolfSSL_SetIO_ISOTP to be called first");
+        return WOLFSSL_CBIO_ERR_TIMEOUT;
+    }
+    isotp_ctx = (struct isotp_wolfssl_ctx*) ctx;
+
+    if (sz > ISOTP_MAX_DATA_SIZE) {
+        WOLFSSL_MSG("ISO-TP packets can be at most 4095 bytes");
+        return WOLFSSL_CBIO_ERR_GENERAL;
+    }
+    /* Can't send whilst we are receiving */
+    if (isotp_ctx->state != ISOTP_CONN_STATE_IDLE) {
+        return WOLFSSL_ERROR_WANT_WRITE;
+    }
+    isotp_ctx->state = ISOTP_CONN_STATE_SENDING;
+
+    /* Assuming normal addressing */
+    if (sz <= ISOTP_SINGLE_FRAME_DATA_SIZE) {
+        ret = isotp_send_single_frame(isotp_ctx, buf, (word16)sz);
+    } else {
+        ret = isotp_send_first_frame(isotp_ctx, buf, (word16)sz);
+    }
+    isotp_ctx->state = ISOTP_CONN_STATE_IDLE;
+
+    if (ret == 0) {
+        return sz;
+    }
+    return ret;
+}
+
+static int isotp_receive_single_frame(struct isotp_wolfssl_ctx *ctx)
+{
+    word16 data_size;
+
+    /* 1 nibble for data size which will be 1 - 7 in a regular 8 byte CAN
+     * packet */
+    data_size = ctx->frame.data[0] & 0xf;
+    if (ctx->receive_buffer_size < data_size) {
```

Comment:
Note: `receive_buffer_size` is `int`. Might consider cast on `data_size` to `int` to avoid unsigned / signed mismatch warnings.